pincer/commands/commands.py

```diff
@@ -168,169 +183,337 @@ async def test_command(
         raise TooManyArguments(
             f"Command `{cmd}` (`{func.__name__}`) can only have 25 "
             f"arguments (excluding the context and self) yet {len(params)} "
             "were provided!"
         )
 
-    options: List[AppCommandOption] = []
-
     for idx, param in enumerate(params):
         if idx == 0 and pass_context:
             continue
 
-        annotation, required = sig[param].annotation, True
+        sig = signature[param]
+
+        annotation, required = sig.annotation, sig.default is _empty
 
         # ctx is type MessageContext but should not be included in the
         # slash command
-        if annotation == MessageContext:
+        if annotation == MessageContext and idx == 1:
             return
 
-        argument_description: Optional[str] = None
-        choices: List[AppCommandOptionChoice] = []
-
-        if isinstance(annotation, str):
-            TypeCache()
-            annotation = eval(annotation, TypeCache.cache, globals())
-
-        if isinstance(annotation, Descripted):
-            argument_description = annotation.description
-            annotation = annotation.key
-
-            if len(argument_description) > 100:
-                raise CommandDescriptionTooLong(
-                    f"Tuple annotation `{annotation}` on parameter "
-                    f"`{param}` in command `{cmd}` (`{func.__name__}`), "
-                    "argument description too long. (maximum length is 100 "
-                    "characters)"
+        if type(annotation) != CommandArg:
+            if annotation in _options_type_link:
+                options.append(
+                    AppCommandOption(
+                        type=_options_type_link[annotation],
+                        name=param,
+                        description="Description not set",
+                        required=required
+                    )
                 )
+                continue
 
-        if get_origin(annotation) is Union:
-            args = get_args(annotation)
-            if type(None) in args:
-                required = False
+            # TODO: Write better exception
+            raise InvalidArgumentAnnotation(
+                "Type must be CommandArg or other valid type"
+            )
 
-            # Do NOT use isinstance as this is a comparison between
-            # two values of the type type and isinstance does NOT
-            # work here.
-            union_args = [t for t in args if t is not type(None)]
+        command_type = _options_type_link[annotation.command_type]
+        argument_description = annotation.get_arg(
+            Description) or "Description not set"
+        choices = annotation.get_arg(
+            Choices)
 
-            annotation = (
-                get_index(union_args, 0)
-                if len(union_args) == 1
-                else Tuple[List]
+        if choices is not MISSING and annotation.command_type not in [int, float, str]:
```

Comment:
if i remember correctly for stuff like membership checking it's better to use sets, correct me if I'm wrong tho
```suggestion
        if choices is not MISSING and annotation.command_type not in {int, float, str}:
```